x-pack/legacy/plugins/apm/public/components/app/ServiceMap/index.tsx

```diff
@@ -37,37 +54,159 @@ ${theme.euiColorLightShade}`,
   margin: `-${theme.gutterTypes.gutterLarge}`
 };
 
+const MAX_REQUESTS = 5;
+
 export function ServiceMap({ serviceName }: ServiceMapProps) {
-  const {
-    urlParams: { start, end }
-  } = useUrlParams();
+  const callApmApi = useCallApmApi();
+  const license = useLicense();
+  const { search } = useLocation();
+  const { urlParams, uiFilters } = useUrlParams();
+  const { notifications } = useApmPluginContext().core;
+  const params = useDeepObjectIdentity({
+    start: urlParams.start,
+    end: urlParams.end,
+    environment: urlParams.environment,
+    serviceName,
+    uiFilters: {
+      ...uiFilters,
+      environment: undefined
+    }
+  });
+
+  const renderedElements = useRef<ElementDefinition[]>([]);
+  const openToast = useRef<string | null>(null);
+
+  const [responses, setResponses] = useState<ServiceMapAPIResponse[]>([]);
+  const [isLoading, setIsLoading] = useState(true);
+  const [percentageLoaded, setPercentageLoaded] = useState(0);
+  const [, _setUnusedState] = useState(false);
```

Comment:
The number of `useState`, `useRef`, `useMemo`, `useEffect` and `useCallback` hooks in this file worries me a little. I'm not saying they are doing something wrong - but it points to a looot of things going on. Would be great if we can find some sensible abstractions that make this easier to comprehend and modify without fear.